clang/lib/Sema/SemaCoroutine.cpp

```diff
@@ -297,6 +297,26 @@ struct ReadySuspendResumeResult {
   bool IsInvalid;
 };
 
+// Adds [[clang::coro_wrapper]] and [[clang::coro_disable_lifetimebound]]
+// attributes to `get_return_object`.
+static void handleGetReturnObject(Sema &S, MemberExpr *ME) {
+  if (!ME || !ME->getMemberDecl() || !ME->getMemberDecl()->getAsFunction())
+    return;
+  auto *MD = ME->getMemberDecl()->getAsFunction();
+  auto *RetType = MD->getReturnType()->getAsRecordDecl();
+  if (!RetType || !RetType->hasAttr<CoroReturnTypeAttr>())
+    return;
+  // `get_return_object` should be allowed to return coro_return_type.
+  if (!MD->hasAttr<CoroWrapperAttr>())
+    MD->addAttr(
+        CoroWrapperAttr::CreateImplicit(S.getASTContext(), MD->getLocation()));
+  // Object arg of `__promise.get_return_object()` is not lifetimebound.
+  if (RetType->hasAttr<CoroLifetimeBoundAttr>() &&
+      !MD->hasAttr<CoroDisableLifetimeBoundAttr>())
+    MD->addAttr(CoroDisableLifetimeBoundAttr::CreateImplicit(
+        S.getASTContext(), MD->getLocation()));
+}
+
```

Comment:
I didn't get this. Can't we just get the decl to `get_return_object` in `CoroutineStmtBuilder::makeReturnObject` and mark it as `CoroWrapperAttr` and `CoroDisableLifetimeBoundAttr`?

Then we should be able to get rid of hard coding the `get_return_object` name in `Sema::CheckCoroutineWrapper` too?

Reply:
Moved the check to `makeReturnObject`.

> Then we should be able to get rid of hard-coding the get_return_object name in Sema::CheckCoroutineWrapper too?

`CheckCoroutineWrapper` is called in `ActOnFinishFunctionBody`. At this point, we are in the `promise::get_return_object` definition. It is possible that we have not seen any coroutine associated with this promise until now. So the coroutine statements would not have been generated.